File: sdk/typescript/src/resources.ts
import { VLogBinClient } from "./client.js";

export interface CreateCustomerInput {
  external_id: string;
  account_type: string;
  display_name: string;
}

export interface Customer {
  id: string;
  external_id: string;
  account_type: string;
  display_name: string;
}

export interface Subscription {
  id: string;
  external_id: string;
  customer_account_id: string;
  catalog_version_id: string;
  plan_id: string;
  status: string;
  started_at: string;
  terminated_at: string | null;
}

export interface IngestUsageInput {
  transaction_id: string;
  customer_external_id: string;
  metric_code: string;
  timestamp: string;
  properties: Record<string, unknown>;
}

export interface Event {
  id: string;
  event_type: string;
  aggregate_id: string;
  transaction_id: string;
  status: string;
  created_at: string;
}

export interface StreamResult {
  events: Event[];
  next_cursor: string | null;
  has_more: boolean;
}

export interface StreamEventsInput {
  cursor?: string;
  limit?: number;
  type?: string;
  aggregateType?: string;
}

export function createCustomer(
  client: VLogBinClient,
  input: CreateCustomerInput,
  idempotencyKey?: string,
): Promise<Customer> {
  return client.request<{ customer: Customer }>(
    "POST",
    "/customers",
    idempotencyKey ? { idempotencyKey } : {},
    input,
  ).then((out) => out.customer);
}

export function listSubscriptions(
  client: VLogBinClient,
): Promise<Subscription[]> {
  return client.request<{ subscriptions: Subscription[] }>(
    "GET",
    "/subscriptions",
  ).then((out) => out.subscriptions);
}

export function ingestUsage(
  client: VLogBinClient,
  input: IngestUsageInput,
  idempotencyKey?: string,
): Promise<{ status: string }> {
  return client.request(
    "POST",
    "/usage/ingest",
    idempotencyKey ? { idempotencyKey } : {},
    input,
  );
}

export function streamEvents(
  client: VLogBinClient,
  input: StreamEventsInput = {},
): Promise<StreamResult> {
  const query: Record<string, string> = { limit: String(input.limit ?? 100) };
  if (input.cursor) query.cursor = input.cursor;
  if (input.type) query.type = input.type;
  if (input.aggregateType) query.aggregate_type = input.aggregateType;
  return client.request("GET", "/events", { query });
}

File: sdk/go/resources.go
package vlogbin

import (
	"context"
	"net/url"
	"strconv"
)

// Customer mirrors the public provider-domain customer resource.
type Customer struct {
	ID          string `json:"id"`
	ExternalID  string `json:"external_id"`
	AccountType string `json:"account_type"`
	DisplayName string `json:"display_name"`
}

// CreateCustomerInput is the public customer creation payload.
type CreateCustomerInput struct {
	ExternalID  string `json:"external_id"`
	AccountType string `json:"account_type"`
	DisplayName string `json:"display_name"`
}

// CreateCustomer creates a customer under the authenticated environment.
func (c *Client) CreateCustomer(ctx context.Context, input CreateCustomerInput) (*Customer, error) {
	var out struct {
		Customer Customer `json:"customer"`
	}
	if err := c.Do(ctx, "POST", "/customers", RequestOptions{}, input, &out); err != nil {
		return nil, err
	}
	return &out.Customer, nil
}

// Subscription mirrors the provider-domain subscription resource.
type Subscription struct {
	ID                string `json:"id"`
	ExternalID        string `json:"external_id"`
	CatalogVersionID  string `json:"catalog_version_id"`
	PlanID            string `json:"plan_id"`
	Status            string `json:"status"`
	CustomerAccountID string `json:"customer_account_id"`
}

// ListSubscriptions lists subscriptions in the authenticated environment.
func (c *Client) ListSubscriptions(ctx context.Context) ([]Subscription, error) {
	var out struct {
		Subscriptions []Subscription `json:"subscriptions"`
	}
	if err := c.Do(ctx, "GET", "/subscriptions", RequestOptions{}, nil, &out); err != nil {
		return nil, err
	}
	return out.Subscriptions, nil
}

// IngestUsageInput is the public usage ingestion payload.
type IngestUsageInput struct {
	TransactionID      string         `json:"transaction_id"`
	CustomerExternalID string         `json:"customer_external_id"`
	MetricCode         string         `json:"metric_code"`
	Timestamp          string         `json:"timestamp"`
	Properties         map[string]any `json:"properties"`
}

// UsageResult is the public ingestion result.
type UsageResult struct {
	Status string `json:"status"`
}

// IngestUsage sends one usage event. Use the same transaction ID for retries.
func (c *Client) IngestUsage(ctx context.Context, input IngestUsageInput) (*UsageResult, error) {
	var out UsageResult
	if err := c.Do(ctx, "POST", "/usage/ingest", RequestOptions{}, input, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Event mirrors the public event stream record.
type Event struct {
	ID            string `json:"id"`
	EventType     string `json:"event_type"`
	AggregateID   string `json:"aggregate_id"`
	TransactionID string `json:"transaction_id"`
	Status        string `json:"status"`
	CreatedAt     string `json:"created_at"`
}

// StreamEventsInput controls event stream pagination.
type StreamEventsInput struct {
	Cursor        string
	Limit         int
	Type          string
	AggregateType string
}

// StreamResult is one page of the event stream.
type StreamResult struct {
	Events     []Event `json:"events"`
	NextCursor string  `json:"next_cursor"`
	HasMore    bool    `json:"has_more"`
}

// StreamEvents pulls one page of events from the cursor-based stream.
func (c *Client) StreamEvents(ctx context.Context, input StreamEventsInput) (*StreamResult, error) {
	query := map[string]string{}
	if input.Cursor != "" {
		query["cursor"] = input.Cursor
	}
	if input.Limit > 0 {
		query["limit"] = strconv.Itoa(input.Limit)
	}
	if input.Type != "" {
		query["type"] = input.Type
	}
	if input.AggregateType != "" {
		query["aggregate_type"] = input.AggregateType
	}
	opts := RequestOptions{Query: url.Values{}}
	for k, v := range query {
		opts.Query.Set(k, v)
	}
	var out StreamResult
	if err := c.Do(ctx, "GET", "/events", opts, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
